pkg/sql/colexec/external/external.go

```diff
@@ -466,22 +475,89 @@ func ReadFile(param *tree.ExternParam, proc *process.Process) (io.ReadCloser, er
 			},
 		},
 	}
+	if param.Parallel {
+		vec.Entries[0].Offset = int64(offset[fileParam.FileIndex-1][0])
+		vec.Entries[0].Size = int64(offset[fileParam.FileIndex-1][1] - offset[fileParam.FileIndex-1][0])
+	}
+	if vec.Entries[0].Size == 0 {
+		return nil, nil
+	}
 	err = fs.Read(param.Ctx, &vec)
 	if err != nil {
 		return nil, err
 	}
 	return r, nil
 }
 
-func getCompressType(param *tree.ExternParam) string {
+func ReadFileOffset(param *tree.ExternParam, proc *process.Process, mcpu int, fileSize int64) ([][2]int, error) {
+	arr := make([][2]int, 0)
+
+	fs, readPath, err := GetForETLWithType(param, param.Filepath)
+	if err != nil {
+		return nil, err
+	}
+	var r io.ReadCloser
+	vec := fileservice.IOVector{
+		FilePath: readPath,
+		Entries: []fileservice.IOEntry{
+			0: {
+				Offset:            0,
+				Size:              -1,
+				ReadCloserForRead: &r,
+			},
+		},
+	}
+	var tailSize []int64
+	var offset []int64
+	for i := 0; i < mcpu; i++ {
+		vec.Entries[0].Offset = int64(i) * (fileSize / int64(mcpu))
+		if err = fs.Read(param.Ctx, &vec); err != nil {
+			return nil, err
+		}
+		r2 := bufio.NewReader(r)
+		line, _ := r2.ReadString('\n')
+		tailSize = append(tailSize, int64(len(line)))
```

Comment:
I think this is wrong, if it turns out to be an unfinished string, you need to move on. As with all csv libraries